From: David Howells <dhowells@redhat.com>
Date: Fri, 30 Aug 2013 16:18:31 +0100
Subject: [PATCH 21/33] X.509: Handle certificates that lack an authorityKeyIdentifier field
Git-commit: 17334cabc814f8847975cddc0e29291af6093464
Patch-mainline: v3.13
References: fate#314574
Target: SLE-12

Handle certificates that lack an authorityKeyIdentifier field by assuming
they're self-signed and checking their signatures against themselves.

Signed-off-by: David Howells <dhowells@redhat.com>
Reviewed-by: Kees Cook <keescook@chromium.org>
Reviewed-by: Josh Boyer <jwboyer@redhat.com>
Acked-by: Lee, Chun-Yi <jlee@suse.com>
---
 crypto/asymmetric_keys/x509_public_key.c |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)

Index: linux-3.12-SLE12/crypto/asymmetric_keys/x509_public_key.c
===================================================================
--- linux-3.12-SLE12.orig/crypto/asymmetric_keys/x509_public_key.c
+++ linux-3.12-SLE12/crypto/asymmetric_keys/x509_public_key.c
@@ -142,8 +142,8 @@ static int x509_key_preparse(struct key_
 		 pkey_algo_name[cert->sig.pkey_algo],
 		 pkey_hash_algo_name[cert->sig.pkey_hash_algo]);
 
-	if (!cert->fingerprint || !cert->authority) {
-		pr_warn("Cert for '%s' must have SubjKeyId and AuthKeyId extensions\n",
+	if (!cert->fingerprint) {
+		pr_warn("Cert for '%s' must have a SubjKeyId extension\n",
 			cert->subject);
 		ret = -EKEYREJECTED;
 		goto error_free_cert;
@@ -152,8 +152,9 @@ static int x509_key_preparse(struct key_
 	cert->pub->algo = pkey_algo[cert->pub->pkey_algo];
 	cert->pub->id_type = PKEY_ID_X509;
 
-	/* Check the signature on the key */
-	if (strcmp(cert->fingerprint, cert->authority) == 0) {
+	/* Check the signature on the key if it appears to be self-signed */
+	if (!cert->authority ||
+	    strcmp(cert->fingerprint, cert->authority) == 0) {
 		ret = x509_check_signature(cert->pub, cert);
 		if (ret < 0)
 			goto error_free_cert;
